{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"代码执行结果.ipynb","provenance":[],"mount_file_id":"10PEsVpfHZ01BqTTsw0VvjBWRYrwy-mYA","authorship_tag":"ABX9TyOXv2n1cBDxoeW26G30twKk"},"kernelspec":{"name":"python3","display_name":"Python 3"}},"cells":[{"cell_type":"markdown","metadata":{"id":"EPCnL5STxSl8"},"source":["# 处理原始数据按cell存\r\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"xykqr2Q-vYge","executionInfo":{"status":"ok","timestamp":1611380816899,"user_tz":-480,"elapsed":220416,"user":{"displayName":"Sicheng Wang","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjGd4TielA7TOD0TBpmvgjXTnlJT_1Tjx2xnN1q=s64","userId":"14545742446481527333"}},"outputId":"46188dbc-f5a7-4667-8bfe-77dfbd2d29eb"},"source":["def define_bins(genome_file, resolutions):\r\n","  '''Takes chromosome sizes, and a list of desired resolutions and defines the\r\n","    bins that cover the genome at the desired resolution(s)'''\r\n","  bins = {}\r\n","  valid_chroms = {}\r\n","  for resolution in resolutions:\r\n","    bins[resolution] = {}\r\n","  for resolution in resolutions:\r\n","    for line in chromsizes:\r\n","      hindex = 0\r\n","      chromname, length = line.split()\r\n","      valid_chroms[chromname] = True\r\n","      for i in range(0, int(length), resolution):\r\n","        bins[resolution][(chromname, i)] = hindex\r\n","        hindex += 1\r\n","  return bins, valid_chroms\r\n","\r\n","def cell_sort(percentages):\r\n","\r\n","  '''Iterate through the percentages file generated by the scHi-C pipeline, and return a list\r\n","      of cell barcodes where the cell has >=1000 unique reads, passes cistrans_cutoff and is 0.95 one species'''\r\n","  cells = {}\r\n","  for line in percentages:\r\n","    hup, mup, huc, muc, val, tot, bc1, bc2, sig, length, ones, twos, threes, fours, ratio, gtype, celltype = line.split()\r\n","    if length == 'Long': continue\r\n","    if sig == 'Randomized': continue\r\n","    if float(hup) >= 0.95 and int(huc) >= 5000 and float(ratio) > 1:\r\n","      identity = ('human', huc, celltype)\r\n","      barcode = bc1 + '-' + bc2\r\n","      cells[barcode] = identity # coverage\r\n","  return cells\r\n","\r\n","def validpair_walk(validpairs, cell_list, resolutions, bins, valid_chroms):\r\n","  '''bedpe_walk walks through a validpairs file, splits out reads to outfiles \r\n","  of the format bin1<t>bin2<t>count<t>norm_count<t>chrom1<t>chrom2'''\r\n","  cell_index = 1\r\n","  cell_matrices = {}\r\n","  for resolution in resolutions:\r\n","    cell_matrices[resolution] = {}\r\n","  for line in validpairs:\r\n","    n1, f1, r1, n2, f2, r2, name, q1, a2, s1, s2, bc1, bc2, frag1, dist1, frag2, dist2 = line.split()\r\n","    species1 = n1.split('_')[0]\r\n","    species2 = n2.split('_')[0]\r\n","    barcode = bc1 + '-' + bc2 \r\n","    if barcode in cell_list:\r\n","      if species1 != species2: continue\r\n","      if species1 != cell_list[barcode][0]: continue\r\n","      if n1 not in valid_chroms: continue\r\n","      if n2 not in valid_chroms: continue\r\n","      for resolution in resolutions:\r\n","        if barcode not in cell_matrices[resolution]:\r\n","          cell_matrices[resolution][barcode] = Counter()\r\n","        pos1_reduce_int = (int(f1)+int(r1))/2 // resolution * resolution\r\n","        pos2_reduce_int = (int(f2)+int(r2))/2 // resolution * resolution\r\n","        bin1 = bins[resolution][(n1, pos1_reduce_int)]\r\n","        bin2 = bins[resolution][(n2, pos2_reduce_int)]\r\n","        if bin1 <= bin2:\r\n","          key = (bin1, bin2, n1, n2)\r\n","        else:\r\n","          key = (bin2, bin1, n2, n1)\r\n","        cell_matrices[resolution][barcode][key] += 1\r\n","  return cell_matrices\r\n","          \r\n","\r\n","def normalizeMatrix(matrix):\r\n","  '''Given a matrix, root normalize the values in the matrix'''\r\n","  cov = Counter()\r\n","  normed = {}\r\n","  for i in matrix:\r\n","    bin1, bin2, chrom1, chrom2 = i\r\n","    count = matrix[i]\r\n","    cov[bin1] += count\r\n","    cov[bin2] += count\r\n","  for i in matrix:\r\n","    bin1, bin2, chrom1, chrom2 = i\r\n","    normed[i] = float(matrix[i]) / sqrt(cov[bin1]) / sqrt(cov[bin2])\r\n","  return normed\r\n","  \r\n","def filter_low_quality_cells(matrices, resolutions):\r\n","\r\n","  min_chrom_contacts = {}\r\n","  for line in chromsizes:\r\n","    chromname, length = line.split()\r\n","    min_chrom_contacts[chromname] = int(length) // 1000000\r\n","  print('min_chrom_contacts:', min_chrom_contacts)\r\n","  del_barcodes = {}\r\n","  for resolution in resolutions:\r\n","    del_barcodes[resolution] = []\r\n","    for barcode in matrices[resolution]:\r\n","      nondiagonal_contacts = 0\r\n","      chrom_contacts = {}\r\n","      for key in matrices[resolution][barcode]:\r\n","        bin1, bin2, chrom1, chrom2 = key\r\n","        count = matrices[resolution][barcode][key]\r\n","        if bin1 != bin2 or chrom1 != chrom2:\r\n","          nondiagonal_contacts += count\r\n","        if chrom1 == chrom2:      \r\n","          if chrom1 not in chrom_contacts:\r\n","            chrom_contacts[chrom1] = count\r\n","          else:\r\n","            chrom_contacts[chrom1] += count\r\n","      high_quality = True\r\n","      if nondiagonal_contacts <= 5000:\r\n","        high_quality = False\r\n","      if len(chrom_contacts) != 23: #破案了，有一个chr22压根没有而不是太少，所以直接漏掉了,所以加这个\r\n","        if barcode == 'GCCTTAGG-TGAGGCAA':\r\n","          print('find you23333')\r\n","        high_quality = False\r\n","      for item in chrom_contacts.items(): \r\n","        chrom, contacts = item\r\n","        if contacts < min_chrom_contacts[chrom]:\r\n","          high_quality = False\r\n","      if high_quality == False:\r\n","        del_barcodes[resolution].append(barcode)\r\n","\r\n","    for barcode in del_barcodes[resolution]:\r\n","      del matrices[resolution][barcode]\r\n","  return matrices\r\n","      \r\n","    \r\n","import os,sys,re\r\n","from collections import Counter\r\n","from math import sqrt\r\n","\r\n","chromsizes = open('/content/drive/MyDrive/毕设/data/Ramani/human_chromsize.txt').readlines()\r\n","percentages = open('/content/drive/MyDrive/毕设/data/Ramani/下载的原始数据集文件/GSM2254215_ML1.percentages.txt').readlines()\r\n","validpairs = open('/content/drive/MyDrive/毕设/data/Ramani/下载的原始数据集文件/GSM2254215_ML1.validPairs.txt').readlines() # 过滤前后474->438\r\n","# percentages = open('/content/drive/MyDrive/毕设/data/Ramani/下载的原始数据集文件/GSM2254217_ML3.percentages.txt').readlines()\r\n","# validpairs = open('/content/drive/MyDrive/毕设/data/Ramani/下载的原始数据集文件/GSM2254217_ML3.validPairs.txt').readlines() # 过滤前后156->55\r\n","resolutions = [1000000]\r\n","\r\n","# 按分辨率切成bin\r\n","bins, valid_chroms = define_bins(chromsizes, resolutions)\r\n","\r\n","# 用percentage文件初步筛选有效的barcode\r\n","cell_list = cell_sort(percentages)\r\n","\r\n","# 用validpairs文件处理细胞信息了\r\n","cell_matrices = validpair_walk(validpairs, cell_list, resolutions, bins, valid_chroms)\r\n","\r\n","print(len(cell_matrices[1000000].items()))\r\n","# 根据论文筛选高质量细胞\r\n","cell_matrices = filter_low_quality_cells(cell_matrices, resolutions)\r\n","print(len(cell_matrices[1000000].items()))\r\n","\r\n","for resolution in resolutions:\r\n","  for barcode in cell_matrices[resolution]:\r\n","    fho_name = \"cell_matrix_data/%s_%s_%s.matrix\" % (cell_list[barcode][2], barcode, resolution)\r\n","    fho = open(fho_name, 'w')\r\n","    norm = normalizeMatrix(cell_matrices[resolution][barcode])\r\n","    for i in norm:\r\n","      fho.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (i[0], i[1], cell_matrices[resolution][barcode][i], norm[i], i[2], i[3]))\r\n","    fho.close() "],"execution_count":15,"outputs":[{"output_type":"stream","text":["474\n","min_chrom_contacts: {'human_chr1': 249, 'human_chr2': 243, 'human_chr3': 198, 'human_chr4': 191, 'human_chr5': 180, 'human_chr6': 171, 'human_chr7': 159, 'human_chr8': 146, 'human_chr9': 141, 'human_chr10': 135, 'human_chr11': 135, 'human_chr12': 133, 'human_chr13': 115, 'human_chr14': 107, 'human_chr15': 102, 'human_chr16': 90, 'human_chr17': 81, 'human_chr18': 78, 'human_chr19': 59, 'human_chr20': 63, 'human_chr21': 48, 'human_chr22': 51, 'human_chrX': 155}\n","find you23333\n","438\n"],"name":"stdout"}]},{"cell_type":"markdown","metadata":{"id":"ZTmPSQBXxYQx"},"source":["# 处理后数据统计与检查\r\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"_L6BvaXMxiVq","executionInfo":{"status":"ok","timestamp":1611380833921,"user_tz":-480,"elapsed":922,"user":{"displayName":"Sicheng Wang","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjGd4TielA7TOD0TBpmvgjXTnlJT_1Tjx2xnN1q=s64","userId":"14545742446481527333"}},"outputId":"18bbf8b1-ea34-462b-b345-5296bd6f5cbb"},"source":["import os\r\n","file_list = os.listdir('./cell_matrix_data')\r\n","\r\n","num_1000000_K562 = 0\r\n","num_1000000_GM12878 = 0\r\n","num_1000000_HeLa = 0\r\n","num_1000000_HAP1 = 0\r\n","\r\n","for name in file_list:\r\n","  if '1000000' in name and 'K562' in name:\r\n","    num_1000000_K562 += 1\r\n","  elif '1000000' in name and 'GM12878' in name:\r\n","    num_1000000_GM12878 += 1\r\n","  elif '1000000' in name and 'HeLa' in name:\r\n","    num_1000000_HeLa += 1\r\n","  elif '1000000' in name and 'HAP1' in name:\r\n","    num_1000000_HAP1 += 1\r\n","  else:\r\n","    continue\r\n","print(len(file_list))\r\n","print(num_1000000_HeLa)\r\n","print(num_1000000_HAP1)\r\n","print(num_1000000_GM12878)\r\n","print(num_1000000_K562)"],"execution_count":16,"outputs":[{"output_type":"stream","text":["493\n","245\n","193\n","12\n","43\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"KkkzneAcwrm0","executionInfo":{"status":"ok","timestamp":1611380895249,"user_tz":-480,"elapsed":900,"user":{"displayName":"Sicheng Wang","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjGd4TielA7TOD0TBpmvgjXTnlJT_1Tjx2xnN1q=s64","userId":"14545742446481527333"}}},"source":["# 读cell_matrix_data文件名得到所有细胞\r\n","import os\r\n","import numpy as np\r\n","\r\n","file_list = os.listdir('./cell_matrix_data')\r\n","f = open('all_cells.txt','w')\r\n","for name in file_list:\r\n","  f.write(name[:-15] + '\\n')\r\n","f.close()"],"execution_count":17,"outputs":[]},{"cell_type":"code","metadata":{"id":"omn-Qe1OySpH","executionInfo":{"status":"ok","timestamp":1611381236991,"user_tz":-480,"elapsed":975,"user":{"displayName":"Sicheng Wang","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GjGd4TielA7TOD0TBpmvgjXTnlJT_1Tjx2xnN1q=s64","userId":"14545742446481527333"}}},"source":["file_list = os.listdir('./cell_matrix_data')\r\n","labels = []\r\n","for name in file_list:\r\n","  labels.append(name.split(sep='_')[0])\r\n","np.save('cell_label.npy', labels)\r\n"],"execution_count":20,"outputs":[]},{"cell_type":"code","metadata":{"id":"3FzzdwvUzrp8"},"source":[""],"execution_count":null,"outputs":[]}]}